pkg/operation/botanist/addons.go

```diff
@@ -443,8 +412,75 @@ func (b *Botanist) generateCoreAddonsChart() (*chartrenderer.RenderedChart, erro
 		"node-problem-detector": common.GenerateAddonConfig(nodeProblemDetector, true),
 		"podsecuritypolicies":   common.GenerateAddonConfig(podSecurityPolicies, true),
 		"shoot-info":            common.GenerateAddonConfig(shootInfo, true),
-		"vpn-shoot":             common.GenerateAddonConfig(vpnShoot, true),
-	})
+	}
+
+	var shootClient = b.K8sShootClient.Client()
+
+	if b.Shoot.KonnectivityTunnelEnabled {
+		konnectivityAgentConfig := map[string]interface{}{
+			"proxyHost": common.GetAPIServerDomain(b.Shoot.InternalClusterDomain),
+			"podAnnotations": map[string]interface{}{
+				"checksum/secret-konnectivity-agent": b.CheckSums["konnectivity-agent"],
+			},
+		}
+
+		// Konnectivity agent related values
+		konnectivityAgent, err := b.InjectShootShootImages(konnectivityAgentConfig, common.KonnectivityAgentImageName)
+		if err != nil {
+			return nil, err
+		}
+
+		values["konnectivity-agent"] = common.GenerateAddonConfig(konnectivityAgent, true)
+
+		// TODO: remove when all Gardener supported versions are >= 1.18 and konnectivity tunnel is the default
+		// Remove vpn-shoot leftovers before switching to konnectivity tunnels
+		secret, err := controller.GetSecretByReference(context.TODO(), shootClient, &corev1.SecretReference{Namespace: metav1.NamespaceSystem, Name: "vpn-shoot"})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+
+		if secret != nil {
+			if err := b.ChartApplierShoot.Delete(context.TODO(), filepath.Join(common.ChartPath, "shoot-core", "components", "charts", "vpn-shoot"), "shoot-core", metav1.NamespaceSystem); err != nil {
+				return nil, err
+			}
+		}
```

Comment:
Hmm, why is this needed? Shouldn't it happen automatically due to the reconciliation loop of the Gardener Resource Manager? I think the only thing that you "manually" have to clean up here is the deletion of the `kube-system/vpn-shoot` `Secret` (because it earlier also got created "manually" here, hence, GRM is unaware of it and won't delete it) - but this code can even be removed in a future version of g/g again (independent of when we drop OpenVPN support at all).